### PR TITLE
Use `scodec.bits.ByteVector` for `Entity.Strict`

### PR DESCRIPTION
**Pull request by armanbilge**
**tl;dr** Because `ByteVector` is specialized for `Byte`, it has idiomatic APIs for byte manipulation and copy-free interop that `Chunk` lacks.

This is a follow-up to my comments in https://github.com/http4s/http4s/issues/5812. See there for more details, summary below:

1. IMHO a `ByteVector` the more idiomatic data structure for, well, bytes. `Chunk[Byte]` is kind of an FS2 thing.
2. `ByteVector` exposes useful ops specifically for byte manipulation, not available on `Chunk[Byte]`. I wanted this in:
    https://github.com/typelevel/feral/pull/214#discussion_r882320272
3. `ByteVector` exposes better APIs for copy-free interop. I wanted this in:
    https://github.com/http4s/http4s-jdk-http-client/pull/644#discussion_r882347645
4. Even though `ByteVector` does not subsume `Chunk`, many common `Chunk`s can become a `ByteVector` without copying.
    https://github.com/typelevel/fs2/blob/79344baae70b6e3d2a3731f5e7f8e3d81ce9eedb/core/shared/src/main/scala/fs2/Chunk.scala#L364-L377
5. Does anyone have an example of a backend where `Chunk` has an advantage over `ByteVector`?

### CODE REVIEW

File: core/shared/src/main/scala/org/http4s/EntityEncoder.scala [hunk 1]
```diff
@@ -141,7 +139,7 @@ object EntityEncoder {
 
   implicit def stringEncoder(implicit charset: Charset = `UTF-8`): EntityEncoder.Pure[String] = {
     val hdr = `Content-Type`(MediaType.text.plain).withCharset(charset)
-    simple(hdr)(s => Chunk.array(s.getBytes(charset.nioCharset)))
+    simple(hdr)(ByteVector.encodeString(_)(charset.nioCharset).toOption.get)
```
Comment (armanbilge) [inline core/shared/src/main/scala/org/http4s/EntityEncoder.scala:142]: So, this is a bit weird. Am I missing something, should this be a `Pure` encoder?

File: core/shared/src/main/scala/org/http4s/Message.scala [hunk 2]
```diff
@@ -768,7 +768,7 @@ object Response extends KleisliSyntax {
   val notFound: Response[Pure] =
     Response(
       Status.NotFound,
-      entity = Entity.Strict(Chunk.array("Not found".getBytes(StandardCharsets.UTF_8))),
+      entity = Entity.Strict(ByteVector.view("Not found".getBytes(StandardCharsets.UTF_8))), // TODO
```
Comment (armanbilge) [inline core/shared/src/main/scala/org/http4s/Message.scala:771]: I opened a PR to add literals e.g. `utf8Bytes"Not found"`.

- https://github.com/scodec/scodec-bits/pull/381

File: core/shared/src/main/scala/org/http4s/EntityEncoder.scala [hunk 3]
```diff
@@ -150,13 +148,13 @@ object EntityEncoder {
...
 
   implicit val chunkEncoder: EntityEncoder.Pure[Chunk[Byte]] =
-    simple(`Content-Type`(MediaType.application.`octet-stream`))(identity)
+    byteVectorEncoder.contramap(_.toByteVector)
 
   implicit val byteArrayEncoder: EntityEncoder.Pure[Array[Byte]] =
-    chunkEncoder.contramap(Chunk.array[Byte])
+    byteVectorEncoder.contramap(ByteVector.view)
 
   implicit def byteVectorEncoder[F[_]]: EntityEncoder[F, ByteVector] =
```
Comment (RafalSumislawski) [inline core/shared/src/main/scala/org/http4s/EntityEncoder.scala:156]: Why is this `EntityEncoder[F, X]` while the others  are `EntityEncoder.Pure[X]`?

**Review by RafalSumislawski — APPROVED** [review]

File: core/shared/src/main/scala/org/http4s/EntityEncoder.scala (hunk 1 above)
Comment (RafalSumislawski) [reply core/shared/src/main/scala/org/http4s/EntityEncoder.scala:142]: I'm slightly inclined towards making it an impure encoder to get rid of this `.toOption.get`.

File: core/shared/src/main/scala/org/http4s/EntityEncoder.scala (hunk 3 above)
Comment (armanbilge) [reply core/shared/src/main/scala/org/http4s/EntityEncoder.scala:156]: Good catch, probably an oversight

File: core/shared/src/main/scala/org/http4s/EntityEncoder.scala [hunk 4]
```diff
@@ -101,14 +101,12 @@ object EntityEncoder {
...
-    encodeBy(hs: _*)(a => Entity.strict(toChunk(a)))
+  def simple[A](hs: Header.ToRaw*)(toByteVector: A => ByteVector): EntityEncoder.Pure[A] =
+    encodeBy(hs: _*)(a => Entity.strict(toByteVector(a)))
 
   /** Encodes a value from its Show instance.  Too broad to be implicit, too useful to not exist. */
-  def showEncoder[A](implicit charset: Charset = `UTF-8`, show: Show[A]): EntityEncoder.Pure[A] = {
-    val hdr = `Content-Type`(MediaType.text.plain).withCharset(charset)
-    simple[A](hdr)(a => Chunk.array(show.show(a).getBytes(charset.nioCharset)))
-  }
+  def showEncoder[A](implicit charset: Charset = `UTF-8`, show: Show[A]): EntityEncoder.Pure[A] =
```
Comment (danicheg) [inline core/shared/src/main/scala/org/http4s/EntityEncoder.scala:108]: This is irrelevant to current PR, but having `Charset` in implicit scope isn't that a good design, I guess. It isn't a typeclass. Should we discuss this before we release 1.0?

File: core/shared/src/main/scala/org/http4s/multipart/Part.scala [hunk 5]
```diff
@@ -44,7 +43,7 @@ object Part {
   def formData(name: String, value: String, headers: Header.ToRaw*): Part[fs2.Pure] =
     Part(
       Headers(`Content-Disposition`("form-data", Map(ci"name" -> name))).put(headers: _*),
-      Entity.Strict(Chunk.array(value.getBytes(UTF_8))),
+      Entity.Strict(ByteVector.encodeUtf8(value).toOption.get),
```
Comment (danicheg) [inline core/shared/src/main/scala/org/http4s/multipart/Part.scala:46]: To work around `.toOption.get` we can do this probably
```suggestion
      Entity.Strict(ByteVector.view(value.getBytes(UTF_8))),
```

**Review by danicheg — COMMENTED** [review]
`scodec` is great and I think we can rely on it. But it's always scary to depend on third-party libraries in the core models :) If `scodec` will decide to break something we will have a classic dilemma.

**Comment by armanbilge** [comment]
> But it's always scary to depend on third-party libraries in the core models :)

scodec-bits is already our dependency in the core models, via fs2 :) itself doesn't have any dependencies and is very stable.

File: core/shared/src/main/scala/org/http4s/EntityEncoder.scala (hunk 4 above)
Comment (armanbilge) [reply core/shared/src/main/scala/org/http4s/EntityEncoder.scala:108]: Yeah definitely worth a discussion. I agree that anything `implicit`-based should be well-justified.

File: core/shared/src/main/scala/org/http4s/multipart/Part.scala (hunk 5 above)
Comment (armanbilge) [reply core/shared/src/main/scala/org/http4s/multipart/Part.scala:46]: @danicheg thanks, I'm pretty curious what the difference is, so some stupid questions: is `value.getBytes(UTF_8)` guaranteed to never throw an error, why is that? What makes it different from `ByteVector.encodeUtf8`? Should this encoder be `Pure` if there can be error?

**Comment by ChristopherDavenport** [comment]
Without digging in too much. Working with Bytevector is awesome, we will be able to do a lot of things with this without getting either into mutable arrays, but will make transformations much simpler.

I'm in favor of this design generally.

File: core/shared/src/main/scala/org/http4s/multipart/Part.scala (hunk 5 above)
Comment (danicheg) [reply core/shared/src/main/scala/org/http4s/multipart/Part.scala:46]: Honestly, I'm not a `scodec` expert, I can just suppose things. 
Speaking from the perspective of the types, we already had the version on `value.getBytes(UTF_8)` and didn't check possible exceptions. Though, I think that `scodec` is much safer if we wouldn't lose on the purpose of this safety through `.toOption.get`. 
Basically, your variant is as safe as the earlier one, I guess. Just users might face a "None.get" exception, instead of what Java API could throw. Will that exception be more clear to users? I dunno.

Comment (armanbilge) [reply core/shared/src/main/scala/org/http4s/multipart/Part.scala:46]: > Basically, your variant is as safe as the earlier one

Thanks, yes, exactly. And the error message is a regression :) since this is 1.0 and we can be breaking things, maybe this should not be `Pure` then.

Comment (armanbilge) [reply core/shared/src/main/scala/org/http4s/multipart/Part.scala:46]: Ah, `getBytes` is total and never throws an error.
> This method always replaces malformed-input and unmappable-character sequences with this charset's default replacement byte array.

**Comment by armanbilge** [comment]
Thanks everyone for the feedback! I think the consensus is overall positive so I will merge this in the next few days unless someone has some new thoughts.

**Comment by armanbilge** [comment]
I've got a PR up optimizing all the `Chunk`-to-`ByteVector` conversions in https://github.com/typelevel/fs2/pull/2956 so they are all copy-free. So in case anybody really needs `Chunk` instead of `ByteVector` it is possible to go back and forth with minimal overhead.

**Comment by rossabaker** [comment]
This is good.

History lesson: scodec-bits was a core dependency all the way back to scalaz-stream days.  fs2-0.9 dropped it, and fs2-0.10 brought it back as a non-core integration.  We never fully got rid of it, but were distancing ourselves from it following fs2's lead.  Once it came back in fs2-core-1.0, there became no particular reason to avoid it, and still lots of reasons above to like it.
